stdlib/public/core/StringUTF8.swift

```diff
@@ -208,27 +208,46 @@ extension String {
     public struct Index : Comparable {
       internal typealias Buffer = _StringCore._UTF8Chunk
 
-      init(_ _core: _StringCore, _ _coreIndex: Int,
-           _ _buffer: Buffer) {
-        self._core = _core
+      init(_ _coreIndex: Int, _ _buffer: Buffer) {
         self._coreIndex = _coreIndex
         self._buffer = _buffer
-        _sanityCheck(_coreIndex >= 0)
-        _sanityCheck(_coreIndex <= _core.count)
+        // FIXME: move sanity checks elsewhere?
+        //_sanityCheck(_coreIndex >= 0)
+        //_sanityCheck(_coreIndex <= _core.count)
       }
 
       /// True iff the index is at the end of its view or if the next
       /// byte begins a new UnicodeScalar.
-      internal var _isOnUnicodeScalarBoundary: Bool {
+      internal func _isOnUnicodeScalarBoundary(of core: _StringCore) -> Bool {
         let buffer = UInt32(truncatingBitPattern: _buffer)
         let (codePoint, _) = UTF8._decodeOne(buffer)
-        return codePoint != nil || _isAtEnd
+        return codePoint != nil || _isAtEnd(of: core)
       }
 
       /// True iff the index is at the end of its view
-      internal var _isAtEnd: Bool {
+      internal func _isAtEnd(of core: _StringCore) -> Bool {
         return _buffer == Index._emptyBuffer
-          && _coreIndex == _core.endIndex
+          && _coreIndex == core.endIndex
+      }
+
+      /// The number of UTF8 code units remaining in the buffer before the next
+      /// `_coreIndex` is reached. This simulates calling `index(after: i)`
+      /// until `i._coreIndex` is incremented, but doesn't need a `_core`
+      /// reference.
+      internal var _codeUnitsBeforeNextCoreIndex: Int {
```

Comment:
Is `_utf8ContinuationBytesUntilNextUnicodeScalar` a better name?

I don't think saying "before next core index" is accurate because a core index can point in the middle of a Unicode scalar, if it takes two UTF-16 code units.